src/test/components/Viewport.test.js

```diff
@@ -0,0 +1,601 @@
+
+      scrollAndGetViewport({
+        // Zoom in some large arbitrary amount:
+        deltaY: -5000,
+        shiftKey: true,
+        clientX: LEFT + COMPONENT_WIDTH * 0.5,
+      });
+
+      // Perform the dragging action some arbitrarily large distance.
+      clickAndDrag(middleX, middleY, middleX - 10000, middleY);
+      expect(getChartProps().viewport.viewportLeft).toBeGreaterThan(0.6);
+      expect(getChartProps().viewport.viewportRight).toBe(1);
+    });
+  });
+});
+
+function setup(profileOverrides: Object = {}) {
+  const flushRafCalls = mockRaf();
+  const ctx = mockCanvasContext();
+
+  jest
+    .spyOn(HTMLCanvasElement.prototype, 'getContext')
+    .mockImplementation(() => ctx);
+
+  jest
+    .spyOn(HTMLElement.prototype, 'getBoundingClientRect')
+    .mockImplementation(() => {
+      const rect = getBoundingBox(COMPONENT_WIDTH, COMPONENT_HEIGHT);
+      // Add some arbitrary X offset.
+      rect.left += LEFT;
+      rect.right += LEFT;
+      rect.x += LEFT;
+      rect.y += TOP;
+      rect.top += TOP;
+      rect.bottom += TOP;
+      return rect;
+    });
+
+  // Hook up a dummy chart with a viewport.
+  const DummyChart = () => <div id="dummy-chart" />;
+  const ChartWithViewport = withChartViewport(DummyChart);
+
+  // This component assumes that it is inside of a connected component, so go
+  // ahead and mock that out here.
+  const ConnectedChartWithViewport = explicitConnect({
+    mapStateToProps: state => ({
+      timeRange: getCommittedRange(state),
+      previewSelection: getPreviewSelection(state),
+    }),
+    mapDispatchToProps: {},
+    // eslint-disable-next-line react/display-name
+    component: (props: Object) => (
+      <ChartWithViewport
+        viewportProps={{
+          previewSelection: props.previewSelection,
+          timeRange: props.timeRange,
+          maxViewportHeight: props.maxViewportHeight || MAX_VIEWPORT_HEIGHT,
+          startsAtBottom: props.startsAtBottom,
+          viewportNeedsUpdate: () => false,
+          marginLeft: 0,
+          marginRight: 0,
+          maximumZoom: MAXIMUM_ZOOM,
+          ...profileOverrides,
+        }}
+        chartProps={
+          {
+            // None used in dummy component.
+          }
+        }
+      />
+    ),
+  });
+
+  const view = mount(
+    <Provider store={storeWithProfile(getProfileFromTextSamples('A').profile)}>
+      <ConnectedChartWithViewport />
+    </Provider>
+  );
+
+  // WithSize uses requestAnimationFrame.
+  flushRafCalls();
+  view.update();
+
+  // The following functions are helpers for the tests, to provide a nicer functional
+  // interface to drive changes to the components.
+
+  function moveMouseAndGetLeft(pageX: number): number {
+    view.simulate('mousemove', { pageX });
+    view.update();
+    return view.find('.timelineTrackScreenshotHover').prop('style').left;
+  }
+
+  function getChartProps() {
+    return view.find(DummyChart).props();
+  }
```

Comment:
everytime you use `getChartProps` in this file, you actually want `viewport`, so maybe the utility function should be `getChartViewportProp`. (optional, because this would involve changing it in the whole file and I know it's a bit tedious).
If you do this, then I believe `scrollAndGetViewport` is actually not so useful, and we could rather just use `scroll` and then `getChartViewportPRop` when we need it.